src/event_manager/manager.rs

```diff
@@ -0,0 +1,415 @@
+use std::sync::{Arc, Mutex};
+
+use crate::epoll::{ControlOperation, EpollEvent, EventSet};
+
+use super::endpoint::EventManagerChannel;
+use super::epoll_context::EpollContext;
+use super::subscribers::Subscribers;
+use super::{
+    ControlOps, Error, EventSubscriber, Events, RemoteEndpoint, Result, SubscriberOps,
+    SubscriberToken,
+};
+
+/// Allows event subscribers to be registered, connected to the event loop, and later removed.
+pub struct EventManager<T> {
+    subscribers: Subscribers<T>,
+    epoll_context: EpollContext,
+    ready_events: Vec<EpollEvent>,
+    channel: EventManagerChannel<T>,
+}
```

Comment:
I'm guessing that this struct is made of generic type `T` so that the users can either do:
- `EventManager<Arc<Mutex<dyn EventSubscriber>>>` - in a multithreaded model
  or
- `EventManager<Rc<RefCell<dyn EventSubscriber>>>`- in a singlethreaded model

right?

I know we can also directly do `EventManager<X: EventSubscriber>`, but that means the `EventManager` takes exclusive ownership of `X` so I don't think this use-case is actually useful.

### Question:
Would it make sense to drop the generic `T` and provide two external interfaces instead of one to the EventManager, one multi-thread-safe `Arc<Mutex<dyn EventSubscriber>>` and one single-thread `Rc<RefCell<dyn EventSubscriber>>`?

**Pros:**
- A single `EventManager` can handle both single-thread and multi-thread objects at the same time. The current generic can't, it's either one or the other.
- The code should become much simpler without the generics.

**Cons:**
- We lose the case `EventManager<X: EventSubscriber>`, but like I said, I'm not sure how useful that is in practice anyway.
- More code, two paths, single and multi-threaded, but should be easy to follow.

What do you think?

Reply:
It will be great if we could remove the generic parameter, there's already a wrapper:
```
impl EventSubscriber for Arc<Mutex<dyn EventSubscriber>>
```

Reply:
@jiangliu What is your concern with using the generic parameter?

I personally don't think we should restrict the interface if we don't have any gain from it. Using `Arc<Mutex<..>>` is not necessary in all VMMs and it isn't a zero cost abstraction. Besides the performance hit that you might get, it is also pretty annoying to use Arc<Mutext<>> because you have to write `.lock().unwrap()` with each usage of the `EventSubscriber`.

In any case, I will get back to you on this topic after I sync with @alexandruag as well.